### PR TITLE
Mulit-model support to ReportManager._add_summary()

### PR DESCRIPTION
**Pull request by nv-braf**
Added multi-model support to ReportManager's _add_summary() method; this is what populates **result_summary.pdf**
I've also updated the formatting/messaging, which will apply to single model runs:

- Updated summary message to include throughput and percentage gain against default model
- Added a page break between the summary plots and the table

The differences for multi-model:

- Print the average latency, as well as each model's latency
- Print the total throughput, as well as each model's throughput
- Print each model's batch size and instance count
- Updated the summary message to include relevant info about both models
- Updated summary plots to use total throughput and average latency

See the picture below for the changes to the message and table for a multi-model run:

Capture1
Capture2

### CODE REVIEW

File: tests/test_report_manager.py [hunk 1]
```diff
@@ -242,16 +248,27 @@ def subtest_build_summary_table(self, mode, cpu_only):
...
+
+        if cpu_only:
+            expected_summary_sentence = (
+                "In 10 measurements across 3 configurations, "
+                "<strong>test_model_config_10</strong> provides the best throughput: <strong>200 infer/sec</strong>.<br><br>"
+                "This is a <strong>200% gain</strong> over the default configuration "
+                "(100 infer/sec), under the given constraints.<UL><LI> "
+                "<strong>test_model_config_10</strong>: 1/GPU model instances with a max batch size of 8 on platform tensorflow_graphdef "
+                "</LI> </UL>")
+        else:
```
Comment (tgerdesnv) [inline tests/test_report_manager.py:263]: Does this test cover all of your refactor and new additions? I think we need more testing here: a high level test like this one for multi-model and/or some lower level tests that hit the lowest level functions to confirm their corner cases.

**Review by tgerdesnv — CHANGES_REQUESTED** [review]

File: model_analyzer/plots/simple_plot.py [hunk 2]
```diff
@@ -87,11 +87,11 @@ def add_run_config_measurement(self, label, run_config_measurement):
                     run_config_measurement.get_gpu_metric_value(
                         tag=self._x_axis))
             else:
+                aggregation_func = sum if self._x_axis == 'perf_throughput' else mean
```
[the remark this answers is not in the document]
Comment (nv-braf) [reply model_analyzer/plots/simple_plot.py]: I've created a tag based static function that returns mean and is overloaded to return sum for perf_throughput

File: model_analyzer/reports/report_manager.py [hunk 3]
```diff
@@ -379,22 +382,16 @@ def _build_summary_report(self, report_key, num_configs, statistics):
                                    image_width=66)
 
         summary.add_paragraph(
+            "<div style = \"display:block; clear:both; page-break-after:always;\"></div>"
             "The following table summarizes each configuration at the measurement"
-            " that optimizes the desired metrics under the given constraints.")
+            " that optimizes the desired metrics under the given constraints."
+            " Per model values are parenthetical.")
```
[the remark this answers is not in the document]
Comment (nv-braf) [reply model_analyzer/reports/report_manager.py:388]: I've removed it for single model.

File: model_analyzer/reports/report_manager.py [hunk 4]
```diff
@@ -324,17 +325,19 @@ def _build_summary_report(self, report_key, num_configs, statistics):
...
             table, summary_sentence = self._build_summary_table(
                 report_key=report_key,
+                num_configurations=total_configurations,
                 num_measurements=total_measurements,
                 cpu_only=True)
 
         # Add summary sections
         summary.add_title(title=f"{self._mode.title()} Result Summary")
-        summary.add_subheading(f"Model: {report_key}")
+        summary.add_subheading(f"Model: {' and '.join(report_key.split(','))}")
```
[the remark this answers is not in the document]
Comment (nv-braf) [reply model_analyzer/reports/report_manager.py:340]: That's correct it's either a comma-separated list of model names or TOP_MODEL_REPORT_KEY, in which case the join is a NOP.

File: model_analyzer/reports/report_manager.py [hunk 5]
```diff
@@ -404,73 +401,311 @@ def _build_summary_table(self,
...
+        return f"{num_measurements} measurements" if num_measurements > 1 else \
+                "1 measurement"
+
+    def _create_summary_config_phrase(self, best_configs, num_configurations):
+        config_names = [
+            f"<strong>{best_config.get_field('name')}</strong>"
+            for best_config in best_configs
+        ]
+
+        config_names_str = f"{' and '.join(config_names)}" if len(config_names) > 1 else \
```
[the remark this answers is not in the document]
Comment (nv-braf) [reply model_analyzer/reports/report_manager.py:472]: You are correct. I've removed the else clause.

File: model_analyzer/reports/report_manager.py [hunk 6]
```diff
@@ -179,33 +178,33 @@ def _add_summary_data(self):
...
-                model_config = result.run_config().model_run_configs(
-                )[0].model_config()
+                model_configs = [
+                    model_run_config.model_config() for model_run_config in
+                    result.run_config().model_run_configs()
+                ]
                 for measurement in result.top_n_measurements(n=1):
                     self._summary_data[model_name].append(
-                        (model_config, measurement))
+                        (model_configs, measurement))
```
[the remark this answers is not in the document]
Comment (nv-braf) [reply model_analyzer/reports/report_manager.py:196]: No good reason. I've replaced all uses of model_configs with run_config

File: model_analyzer/reports/report_manager.py [hunk 7]
```diff
@@ -404,73 +401,311 @@ def _build_summary_table(self,
         model
         """
 
-        if not cpu_only:
-            summary_table = ResultTable(headers=[
+        best_configs, best_run_config_measurement, sorted_measurements, model_config_dicts = self._find_best_configs(
```
[the remark this answers is not in the document]
Comment (nv-braf) [reply model_analyzer/reports/report_manager.py:404]: Replacing model_configs with run_config removed the need for the dict. I now just return the best and sort_measurements. I think it's clean enough now.

File: model_analyzer/reports/report_manager.py [hunk 8]
```diff
@@ -404,73 +401,311 @@ def _build_summary_table(self,
...
+    def _create_summary_max_batch_size_phrase(self, best_configs):
+        max_batch_sizes = [
+            str(best_config.max_batch_size()) for best_config in best_configs
+        ]
+
+        return f"max batch sizes of \"{', '.join(max_batch_sizes)}\"" if len(max_batch_sizes) > 1 \
+          else f"max batch size of {max_batch_sizes[0]}"
+
+    def _create_summary_dynamic_batching_phrase(self, best_configs):
+        # Dynamic batching is either on/off for all model configs in a run config
```
[the remark this answers is not in the document]
Comment (nv-braf) [reply model_analyzer/reports/report_manager.py:517]: Point taken. I'm now printing all models dynamic batching strings

File: tests/test_report_manager.py (hunk 1 above)
Comment (nv-braf) [reply tests/test_report_manager.py:263]: I agree that we need more testing here, but I don't think our current approach gives us the best !/$, it's just not scalable for multi-model. I've filed [TMA-666](https://jirasw.nvidia.com/browse/TMA-666) to add checkpoint based unit testing.

For single model, we have decent coverage with the current unit tests. And, I've got the multi-model correct by inspection (running analysis and looking at the report pdf).

File: model_analyzer/record/record.py [hunk 9]
```diff
@@ -121,6 +122,18 @@ def aggregation_function():
 
         return (lambda records: max(records, key=lambda r: r.value()))
 
+    @staticmethod
+    def value_function():
+        """
+        The function that is used to combine records from multiple
```
Comment (tgerdesnv) [inline model_analyzer/record/record.py:128]: We need clear description to distinguish between aggregation_function. what is different?  That one shows that it takes a list of records and returns a single record.  This takes a list of records(?) and returns a float/int?

File: model_analyzer/plots/simple_plot.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tgerdesnv) [reply model_analyzer/plots/simple_plot.py]: Nice!

File: model_analyzer/plots/simple_plot.py [hunk 10]
```diff
@@ -14,6 +14,7 @@
 
 import os
 import matplotlib.pyplot as plt
+from statistics import mean
```
Comment (tgerdesnv) [inline model_analyzer/plots/simple_plot.py:17]: no longer needed

File: model_analyzer/reports/report_manager.py [hunk 11]
```diff
@@ -20,6 +20,7 @@
 from .report_factory import ReportFactory
 
 import os
+from statistics import mean
```
Comment (tgerdesnv) [inline model_analyzer/reports/report_manager.py:23]: no longer needed

File: model_analyzer/reports/report_manager.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tgerdesnv) [reply model_analyzer/reports/report_manager.py:517]: Are you? I don't see changes. I'm fine with it printing just enabled or just disabled if they are all the same (the most common case), but we need the ability to split it up if different.

**Review by tgerdesnv — CHANGES_REQUESTED** [review]

File: model_analyzer/plots/simple_plot.py (hunk 10 above)
Comment (nv-braf) [reply model_analyzer/plots/simple_plot.py:17]: Removed

File: model_analyzer/reports/report_manager.py (hunk 11 above)
Comment (nv-braf) [reply model_analyzer/reports/report_manager.py:23]: Removed

File: model_analyzer/record/record.py (hunk 9 above)
Comment (nv-braf) [reply model_analyzer/record/record.py]: Updated the comment and used the lambda function

File: model_analyzer/reports/report_manager.py [hunk 12]
```diff
@@ -474,16 +474,17 @@ def _create_summary_throughput_phrase(self, report_key,
...
+            100) if default_throughput else None
 
         throughput_phrase = "total " if multi_model else ""
         throughput_phrase = (
             throughput_phrase +
             f"throughput: <strong>{best_throughput} infer/sec</strong>.<br><br>"
         )
 
-        if throughput_gain:
+        if throughput_gain is not None:
```
Comment (tgerdesnv) [inline model_analyzer/reports/report_manager.py:487]: Can we also exclude if throughput_gain is 0?

**Review by tgerdesnv — CHANGES_REQUESTED** [review]

**Review by tgerdesnv — APPROVED** [review]
